implement general data model - agnostic starting point for further development

magland opened the pull request:
@jsoules @WardBrian 

With this PR I am abstracting out the analysis data model. (An analysis for now is something that has stan file content, data file content, and sampling opts file content.)

There is no sharing functionality nor loading from the cloud in this branch. So this is agnostic to the method we end up using for that.

This is a starting point (base layer) for tryout out different ways to implement that.

Let me know what you think.

magland commented:
.... also, it does not have any functionality for storing local changes.

WardBrian commented:
> .... also, it does not have any functionality for storing local changes.

Does this mean refreshing the page loses everything?

magland commented:
> > .... also, it does not have any functionality for storing local changes.
> 
> Does this mean refreshing the page loses everything?

Yes, but that's just until we settle on the right way to be saving local edits.

jsoules reviewed (commented):
I think this makes some good changes, but I also have a few things I'd like to hash out a bit further.

I'd like to understand better how we decide what data lives in the context vs. in state on the HomePage component. Why not have most data dependencies in the front-end be dependent on the context state?

Similarly, I think it would be clearer to code against the data model if the list of expected/required files were more definite. The `SPAnalysisDataModel` clearly embeds some assumptions about what files to expect, but the `AnalysisFiles` and corresponding reducer don't match up or enforce them.

Finally, while I know we're still undecided about the mix-and-match vs all-in-one model for sharing and persisting an `Analysis`, it would be nice to keep using local storage for changes in the meantime, assuming that's possible without setting strong expectations about those parts of the design.

Anyway, have a nice weekend and looking forward to a design discussion on Monday!

magland commented on gui/src/app/SPAnalysis/SPAnalysisContext.ts:
```diff
@@ -0,0 +1,25 @@
+
+export type SPAnalysisDataModel = {
+    title: string
+    stanFileContent: string
+    setStanFileContent: (text: string) => void
+    dataFileContent: string
+    setDataFileContent: (text: string) => void
+    samplingOptsContent: string
+    setSamplingOptsContent: (text: string) => void
```
Noted. At this point, we only support one of each, and that's appropriately reflected here.

magland commented on gui/src/app/SPAnalysis/SetupSPAnalysis.tsx:
```diff
@@ -0,0 +1,77 @@
+import { SPAnalysisContext } from "./SPAnalysisContext"
+
+type SetupSPAnalysisProps = {
+    sourceDataUri: string
```
Yes that's right. I have added a small comment in the source to indicate that.

magland commented on gui/src/app/SPAnalysis/SetupSPAnalysis.tsx:
```diff
@@ -0,0 +1,77 @@
+}
+
+const analysisFilesReducer = (state: AnalysisFiles, action: AnalysisFilesAction): AnalysisFiles => {
+    switch (action.type) {
+        case 'set': {
+            return {
+                ...state,
+                [action.key]: action.value
+            }
+        }
+        case 'delete': {
+            const newState = { ...state }
+            delete newState[action.key]
+            return newState
+        }
+    }
```
Yeah this is internal implementation. I have renamed this from AnalysisFiles to KVStore to suggest that it is just a key-value store used internally for the implementation.

magland commented on gui/src/app/SPAnalysis/SetupSPAnalysis.tsx:
```diff
@@ -0,0 +1,77 @@
+    const value = useMemo(() => {
+        return {
+            localDataModel: {
+                title: 'SPAnalysis',
```
I have changed this to 'Untitled' and added the comment:

"// title is hard-coded for now because we don't yet have a mechanism for it to be changed"

magland commented on gui/src/app/SPAnalysis/SetupSPAnalysis.tsx:
```diff
@@ -0,0 +1,77 @@
+
+type AnalysisFiles = {
+    [key: string]: string
+}
```
As noted above, I have renamed this from AnalysisFiles to KVStore to suggest that it is just a key-value store used internally for the implementation. It is left general so it can be replaced by something more specific.

magland commented on gui/src/app/pages/HomePage/HomePage.tsx:
```diff
@@ -1,139 +1,134 @@
-const defaultSamplingOptsContent = JSON.stringify(defaultSamplingOpts)
+const HomePage: FunctionComponent<Props> = ({ width, height }) => {
+    const { route } = useRoute()
+    if (route.page !== 'home') {
+        throw Error('Unexpected route')
+    }
```
Agreed

magland commented on gui/src/app/pages/HomePage/HomePage.tsx:
```diff
@@ -1,139 +1,134 @@
-            return
+    useEffect(() => {
+        // update the document title based on the route
+        if (!route.title) {
```
I changed it to your suggestion.

magland commented on gui/src/app/pages/HomePage/HomePage.tsx:
```diff
@@ -1,139 +1,134 @@
+    ), [localDataModel])
+    const saveDataFileContent = useCallback((text: string) => {
+        localDataModel?.setDataFileContent(text)
+    }, [localDataModel])
```
You're right. I have removed the unnecessary logic and for example now use `localDataModel.setStanFileContent` directly in place of `saveStanFileContent`. However editedStanFileContent state still needs to exist separately and be in this HomPage component (and note the context). This is text that is edited in the editor but not yet saved.

jsoules replied:
Okay. There's a part of me that wants to put this in the `context` anyway, but if (as I think is currently the case by definition) the edited/unsaved code should not be visible outside this component, I agree this makes sense.

I could see a model like this getting too heavy-weight for what's in the component if we wind up doing the same thing for multiple other files (i.e. data.py, analysis.py) If we do go that route, I would want to put all the editor-pane contents into the context, possibly with corresponding "unsaved" flags; or even move to a no-explicit-saving paradigm.

But we can worry about that when we get there.

magland commented on gui/src/app/SPAnalysis/SPAnalysisContext.ts:
```diff
@@ -0,0 +1,25 @@
+import { createContext, useContext } from "react"
+
+export type SPAnalysisDataModel = {
```
I have added a small comment in the source to this effect.

magland commented on gui/src/app/pages/HomePage/HomePage.tsx:
```diff
@@ -1,139 +1,134 @@
+        if (!localDataModel) return
+        const newRoute = { ...route, title: localDataModel.title }
+        setRoute(newRoute, true)
+    }, [localDataModel.title, route, setRoute])
```
I have removed that falsy check, and it turns out it wasn't necessary. Now linter is happy.

magland commented:
> I think this makes some good changes, but I also have a few things I'd like to hash out a bit further.

Thanks for the helpful review @jsoules. I have responded to each of your points above.


> I'd like to understand better how we decide what data lives in the context vs. in state on the HomePage component. Why not have most data dependencies in the front-end be dependent on the context state?

I have removed the unnecessary extra state in the HomePage as indicated above. The edited (unsaved) state does need to stay in HomePage because that's going to be handled locally for that component.

> Similarly, I think it would be clearer to code against the data model if the list of expected/required files were more definite. The `SPAnalysisDataModel` clearly embeds some assumptions about what files to expect, but the `AnalysisFiles` and corresponding reducer don't match up or enforce them.

As indicated above, I have renamed AnalysisFiles to KVStore to indicate that this is an internal implementation that could be swapped out.

> 
> Finally, while I know we're still undecided about the mix-and-match vs all-in-one model for sharing and persisting an `Analysis`, it would be nice to keep using local storage for changes in the meantime, assuming that's possible without setting strong expectations about those parts of the design.

Okay. I have restored this functionality for convenience during development. And I have added the following comment

```
// For convenience, we save the state to local storage so it is available on
// reload of the page But this will be revised in the future to use a more
// sophisticated storage mechanism.
```

> 
> Anyway, have a nice weekend and looking forward to a design discussion on Monday!

Thanks, you too!
